cincinnati/src/plugins/internal/arch_filter.rs

```diff
@@ -0,0 +1,356 @@
+                    semver::Version::from_str(&default_arch_threshold_version)
+                        .map_err(|e| GraphError::ArchVersionError(e.to_string()))?,
+                )
+            };
+
+            if semantic_version > threshold_semantic_version {
+                return Err(GraphError::MissingParams(
+                    vec!["arch"].into_iter().map(String::from).collect(),
+                ));
+            }
+
+            debug!(
+                "no architecture given. inferring {} by version {} <= {}",
+                default_arch, version, default_arch_threshold_version
+            );
+
+            Ok(default_arch)
+        }
+        (None, None) => Err(GraphError::MissingParams(
+            vec!["arch", "version"]
+                .into_iter()
+                .map(String::from)
+                .collect(),
+        )),
+    }
+}
+
+/// Regex for arch label validation.
+static ARCH_VALIDATION_REGEX_STR: &str = r"^[0-9a-z]+$";
+
+lazy_static! {
+    static ref ARCH_VALIDATION_REGEX_RE: regex::Regex =
+        regex::Regex::new(&ARCH_VALIDATION_REGEX_STR).expect("could not create regex");
+}
+
+impl InternalPlugin for ArchFilterPlugin {
+    fn run_internal(self: &Self, internal_io: InternalIO) -> AsyncIO<InternalIO> {
+        let arch = {
+            let interesting_params =
+                try_get_multiple_values!(internal_io.parameters, "arch", "version");
+
+            infer_version(
+                interesting_params.0.map(std::string::ToString::to_string),
+                interesting_params.1.map(std::string::ToString::to_string),
+                self.default_arch.clone(),
+                self.default_arch_threshold_version.clone(),
+            )
+        }
+        .map_err(failure::Error::from);
+
+        let future_result = futures::future::result(arch)
+            .join(futures::future::ok::<_, failure::Error>((
+                internal_io,
+                self.key_prefix.to_owned(),
+                self.key_suffix.to_owned(),
+            )))
+            .and_then(|(arch, (internal_io, key_prefix, key_suffix))| {
+                let mut graph = internal_io.graph;
+
+                // iterate over all releases which have the arch metadata set and
+                // 1. for every release which matches the given `arch`, rewrite the
+                //    arch metadata to only contain `arch`
+                // 2. collect all other releases to be removed
+                let to_remove = {
+                    graph
+                        .find_by_fn_mut(|release| {
+                            match release {
+                                crate::Release::Concrete(concrete_release) => concrete_release
+                                    .metadata
+                                    .get_mut(&format!("{}.{}", key_prefix, key_suffix))
+                                    .map_or(true, |values| {
+                                        if values.split(',').any(|value| value.trim() == arch) {
```

Comment:
This should be looking at the arch extracted from the release image via camallo/dkregistry-rs#117, right?

Reply:
> Or is the graph builder using the dkregistry-rs code to fill in the arch metadata being consumed here?

Yes